# <center> Panongbene Jean Mohamed Sawadogo </center>#

* $\color{red}{Course}$        : Visual Analytics, CentraleSupelec
* $\color{red}{Professor}$     : Petra Isenberg, Wesley Willett
* $\color{red}{Assignement}$   : second Assignement
* $\color{red}{Author}$        : Panongbene Jean Mohamed Sawadogo
* $\color{red}{Title}$         : Collecting & Cleaning Data

$\color{red}{Project - description : }$  In this project, we are scraping the site https://www.populationdata.net/pays/ to extract all informations relating to the countries of the world at the end, we will build a csv containing the countries with as much information as possible about each country.





# We import the necessary libraries

In [1]:
import pandas as pd
from bs4 import BeautifulSoup
import requests 

# <center> Part 1 : Scraping Data From the Web </center> #
In this part, we build and declare all the functions and variables needed to retrieve data from the website.

# All Extras Parameters

In [2]:
link_web_page = 'https://www.populationdata.net/pays/'

# the real name in the web page of the dataframe columns
array_column = ['Nom officiel','Nom propre','Continent','Sous-continent','Population',
                'Croissance démographique','Superficie','Densité','PIB','PIB/habitant',
                'Croissance du PIB','Espérance de vie',
                'Taux de natalité','Indice de fécondité','Taux de mortalité','Taux de mortalité infantile(‰ (2019))',
                'Taux d\'alphabétisation','Langues officielles','Monnaie','IDH','IPE','Nature de l"État','Chef de l"État',
                'Fête nationale','Codes ISO','Gentilé','Touristes']

# the names of all datafram columns
datafram_column = ['Nom officiel','Nom propre','Continent','Sous-continent','Population(habitants (2020))',
                  'Croissance démographique(% / an)','Superficie(km²)','Densité(habitants / km²)',
                  'PIB(milliards $USD (2017))','PIB/habitant','Croissance du PIB(% / an (2018))',
                  'Espérance de vie(ans (2019))','Taux de natalité(‰ (2019))','Indice de fécondité(enfants / femme (2019))',
                  'Taux de mortalité(‰ (2019))','Taux de mortalité infantile(‰ (2019))','Taux d"alphabétisation(% (2020))',
                   'Langues officielles','Monnaie','IDH(/ 1 (2018))','IPE','Nature de l"État','Chef de l"État','Fête nationale',
                  'Codes ISO','Gentilé','Touristes']

In [3]:
def add_new_line_data_fram(name, dictionnary_data_new,datafram_country_data):
    """Use this function to add a new row in our data frame. The new row values ​​are
    the values ​​in dictionnary_data_new"""
    
    # We create the new line in data Fram
    datafram_country_data.loc[name] = 'NAN'
    
    # We put the values in our new line
    j = 0
    for i in dictionnary_data_new:
        datafram_country_data.loc[name][j] = dictionnary_data_new[i]
        j+=1

def save_data_frame_csv(file_name, dataframe_data):
    """We use this function to save dataframe in the csv file"""
    my_file_name = file_name + '.csv'
    dataframe_data.to_csv(my_file_name)
    print("Saved " +my_file_name)

## Function to Scrape and analyze data


In [4]:
#function 
def fetch_scrapping_pages():
    """We use this function to scrapes all informations about each country. 
     We use also this function for cleaning the data in the datafram. At the 
     end this function return a panda dataframe which contains all cleaning informations 
     of each countries. """
    
    # page Request
    page_country_first = requests.get(link_web_page, allow_redirects=True)
    
    #we process the request
    if not page_country_first.status_code == 200:
        print(" ERROR Status code: " + str(page_country_first.status_code))
        return
    else:
        soup = BeautifulSoup(page_country_first.content)
        
        # we create the data frame
        datafram_country_data = pd.DataFrame(columns=datafram_column)
    
        for tablecountry in soup.findAll('table'):
            for onecountry in tablecountry.findAll('tr'):
                # name country
                name_country = ''
                # we create the dictionnary
                dictionnary_data = {}
                
                if('data-iso-code' in onecountry.attrs):
                    for country in onecountry.findAll('a'):
                        if('pays' in country['href']):
                            link_country = country['href']
                            name_country = country.text
                            
                            # page Request
                            page_country = requests.get(link_country, allow_redirects=True)
                            #we process the request
                            if not page_country.status_code == 200:
                                print(" ERROR Status code: " + str(page_country.status_code))
                                continue
                            else:
                                soup_page_country = BeautifulSoup(page_country.content)
                                
                                for tablecountry in soup_page_country.findAll('tbody'):
                                    for table_infos in tablecountry.findAll('tr'):
                                        keys_values = []
                                        for information_line in table_infos.findAll('td'):
                                            keys_values.append(information_line.text)
                                       
                                        #We remove all values ; because ; is interpreted by csv as a column separator
                                        keys_values[0] = keys_values[0].replace(";","")
                                        keys_values[1] = keys_values[1].replace(";","")
                                        
                                        #we replace all nc value by NaN
                                        keys_values[0] = keys_values[0].replace("n.c.","NaN")
                                        keys_values[1] = keys_values[1].replace("n.c.","NaN")

                                        dictionnary_data[keys_values[0]] = keys_values[1]
                                    #we use break here because some page have a lot of tbody
                                    break
                                # we add the new ligne in the datafram
                                add_new_line_data_fram(name_country,dictionnary_data,datafram_country_data)
    return datafram_country_data

In [5]:
datafram_country_data = fetch_scrapping_pages()
datafram_country_data.head()

Nom officiel  \
Abkhazie                     République d'Abkhazie   
Açores                  Région autonome des Açores   
Afghanistan     République islamique d'Afghanistan   
Afrique du Sud         République d'Afrique du Sud   
Albanie                       République d'Albanie   

                                                       Nom propre Continent  \
Abkhazie        Аҧсны Аҳәынҭқарра (ab)  აფხაზეთი (ka)  Республ...      Asie   
Açores                            Região Autónoma dos Açores (pt)    Europe   
Afghanistan     د أفغانستان اسلامي جمهوریت (ps)  جمهوری اسلامی...      Asie   
Afrique du Sud  Republiek van Suid-Afrika (af)  Republic of So...   Afrique   
Albanie                               Republika e Shqipërisë (sq)    Europe   

                                Sous-continent Population(habitants (2020))  \
Abkhazie                       Asie de l'Ouest     246 313 habitants (2018)   
Açores                        Union Européenne     243 862 habitants (2018)   
Afghanistan                      Asie centrale  34 233 212 habitants (2020)   
Afrique du Sud                Afrique Australe  57 725 606 habitants (2018)   
Albanie         Europe (hors Union Européenne)   2 870 324 habitants (2018)   

               Croissance démographique(% / an) Superficie(km²)  \
Abkhazie                            0,49 % / an       8 653 km²   
Açores                             -0,39 % / an       2 322 km²   
Afghanistan                         6,23 % / an     652 864 km²   
Afrique du Sud                      2,13 % / an   1 219 912 km²   
Albanie                            -0,22 % / an      28 748 km²   

               Densité(habitants / km²)     PIB(milliards $USD (2017))  \
Abkhazie          28,47 habitants / km²    0,336 milliards $USD (2017)   
Açores           105,02 habitants / km²    4,220 milliards $USD (2015)   
Afghanistan       52,44 habitants / km²   19,363 milliards $USD (2018)   
Afrique du Sud    47,32 habitants / km²  366,298 milliards $USD (2018)   
Albanie           99,84 habitants / km²   15,059 milliards $USD (2018)   

                      PIB/habitant  ...  \
Abkhazie         1 372 $USD (2017)  ...   
Açores          17 398 $USD (2015)  ...   
Afghanistan        521 $USD (2018)  ...   
Afrique du Sud   6 340 $USD (2018)  ...   
Albanie          5 254 $USD (2018)  ...   

                                    Langues officielles  \
Abkhazie                       Abkhaze, géorgien, russe   
Açores                                        Portugais   
Afghanistan                      Dari (persan), pachtou   
Afrique du Sud  Zoulou, afrikaans, anglais (11 langues)   
Albanie                                        Albanais   

                                            Monnaie   IDH(/ 1 (2018))  \
Abkhazie        Apsar abkhaze, Rouble russe (₽ RUB)               NaN   
Açores                                 Euro (€ EUR)               NaN   
Afghanistan                           Afghani (AFN)  0,496 / 1 (2018)   
Afrique du Sud                           Rand (ZAR)  0,705 / 1 (2018)   
Albanie                                   Lek (ALL)  0,791 / 1 (2018)   

                         IPE  \
Abkhazie                 NaN   
Açores                   NaN   
Afghanistan     37,74 (2018)   
Afrique du Sud  44,73 (2018)   
Albanie         65,46 (2018)   

                                                 Nature de l"État  \
Abkhazie                                               République   
Açores                                                 République   
Afghanistan                                  République islamique   
Afrique du Sud  République unitaire parlementaire, régime prés...   
Albanie                                                République   

                                   Chef de l"État  \
Abkhazie                 Président Raul Khadjimba   
Açores          Président Marcelo Rebelo de Sousa   
Afghanistan                Président Ashraf Ghani   
Afrique du Sud          Président Cyril Ra

In [6]:
# we save the data int the local repertory
save_data_frame_csv('Lab2-country_Information_scrapping', datafram_country_data)

Saved Lab2-country_Information_scrapping.csv


# <center> Part 2 : Cleaining data </center> #
In this part, we explain how we use Openrefine for cleaning data

<span style="color:red"><font size="+1"> INTRODUCTION : </font></span> After scraping the data and saving it to a csv file, we will clean the data using OpenRefine.
We must clean the data because if we take for example the population column, we expect to have a numerical value but, we end up with values of this type <span style="color:green">57 725 606 habitants (2018)</span>.
In this part, we'll see how we use OpenRefine to clean up data.

<span style="color:red"><font size="+1"> Step 1 Install and run openRefine : </font></span> We download, unzip openrif software in this website  https://openrefine.org/download.html. We run OpenRefine binary program with this command <span style="color:green">./refine </span>
<img src="imageExplainOpenRefine/1.png"/>
<img src="imageExplainOpenRefine/2.png"/>
After we run openrefine, the software open a webserver in a local host using the port 3333 <span style="color:green">(127.0.0.1:3333)</span>


<span style="color:red"><font size="+1"> Step 2 Load the Data : </font></span> We load the csv file we have that we built in part one and we create a new project.
<img src="imageExplainOpenRefine/3.png"/>


<span style="color:red"><font size="+1"> Step 3 Cleaning the population column : </font></span> We divided the population column into two columns, the first column contains the population number and the second column contains the date of the census
<img src="imageExplainOpenRefine/4.png"/>
Before cleaning the column

<img src="imageExplainOpenRefine/5.png"/>
After cleaning the population column



<span style="color:red"><font size="+1"> Step 4 Cleaning the Croissance démographique, Superficie and Densité columns : </font></span> 
We transformed the values of the Croissance démographique, Superficie and Densité columns into numerical values
    
<img src="imageExplainOpenRefine/6.png"/>
Before cleaning the column

<img src="imageExplainOpenRefine/7.png"/>
After cleaning the population column


<span style="color:red"><font size="+1"> Step 5 Clean the last columns : </font></span> 
For cleaning up the remaining columns, for some we have divided each column into two columns: the first contains the numeric value of the column and the second the date corresponding to that value. We've also changed the names of some columns to make them more meaningful.


<img src="imageExplainOpenRefine/8.png"/>
Before cleaning the column

<img src="imageExplainOpenRefine/9.png"/>
After cleaning the population column



<span style="color:red"><font size="+1"> Step 6 Finish the job : </font></span> For this step, we replaced the empty cells with NaN and we managed the cells that did not pass the filters individually.

# The head of the file Result of OpenRefine Cleaning

In [7]:
#data_openrefine_result = pd.read_csv("Lab2-country_Information_scrapping-result-openrefine-operation.xlsx")
data_openrefine_result = pd.read_excel("Lab2-country_Information_scrapping-result-openrefine-operation.xlsx")

#pd.read_excel("the_document.ods", engine="odf")

In [8]:
data_openrefine_result

Column  \
0                                             Abkhazie   
1                                               Açores   
2                                          Afghanistan   
3                                       Afrique du Sud   
4                                              Albanie   
5                                              Algérie   
6                                            Allemagne   
7                                              Andorre   
8                                               Angola   
9                                             Anguilla   
10                                         Antarctique   
11                                  Antigua-et-Barbuda   
12                              Antilles néerlandaises   
13                                     Arabie saoudite   
14                                           Argentine   
15                                             Arménie   
16                             Artsakh (Haut-Karabakh)   
17                                               Aruba   
18                                           Australie   
19                                            Autriche   
20                                         Azerbaïdjan   
21                                             Bahamas   
22                                             Bahreïn   
23                                          Bangladesh   
24                                             Barbade   
25                               Bélarus (Biélorussie)   
26                                            Belgique   
27                                              Belize   
28                                               Bénin   
29                                            Bermudes   
..                                                 ...   
223                                           Suriname   
224                                              Syrie   
225                                        Tadjikistan   
226                                             Taïwan   
227                                           Tanzanie   
228                                              Tchad   
229                                           Tchéquie   
230  Terres australes et antarctiques françaises (T...   
231                                          Thaïlande   
232                                     Timor oriental   
233                                               Togo   
234                                            Tokelau   
235                                              Tonga   
236                                       Transnistrie   
237                                  Trinité-et-Tobago   
238                                            Tunisie   
239                                       Turkménistan   
240                                 Turques-et-Caïques   
241                                            Turquie   
242                                             Tuvalu   
243                                            Ukraine   
244                                            Uruguay   
245                                            Vanuatu   
246                                            Vatican   
247                                          Venezuela   
248                                           Viêt Nam   
249                                   Wallis-et-Futuna   
250                                              Yémen   
251                                             Zambie   
252                                           Zimbabwé   

                                        Nom officiel  \
0                              République d'Abkhazie   
1                         Région autonome des Açores   
2                 République islamique d'Afghanistan   
3                        République d'Afrique du Sud   
4                               République d'Albanie   
5    République algérienne démocratique et populaire   
6                    République Fédérale d'Allemagne   
7  

# Json file result of OpenRefine Cleaning

In [9]:
openrefine_history_result = pd.read_json("Lab2-historyOpenRefineScrapping.json")

In [10]:
openrefine_history_result

columnName  \
0                        Population(habitants (2020))   
1                      Population(habitants (2020)) 2   
2                                                 NaN   
3                    Croissance démographique(% / an)   
4                                     Superficie(km²)   
5                            Densité(habitants / km²)   
6                          PIB(milliards $USD (2017))   
7                        PIB(milliards $USD (2017)) 2   
8                                        PIB/habitant   
9                                      PIB/habitant 2   
10                   Croissance du PIB(% / an (2018))   
11                 Croissance du PIB(% / an (2018)) 2   
12                       Espérance de vie(ans (2019))   
13                     Espérance de vie(ans (2019)) 2   
14                         Taux de natalité(‰ (2019))   
15                       Taux de natalité(‰ (2019)) 2   
16        Indice de fécondité(enfants / femme (2019))   
17      Indice de fécondité(enfants / femme (2019)) 2   
18                                                NaN   
19                                                NaN   
20                                                NaN   
21                                                NaN   
22                                                NaN   
23                                                NaN   
24                                                NaN   
25                                                NaN   
26                                                NaN   
27                                                NaN   
28                                                NaN   
29                                                NaN   
..                                                ...   
41                                                NaN   
42                                                NaN   
43                                                IPE   
44                                                NaN   
45                                                NaN   
46                                          Touristes   
47                                        Touristes 2   
48                                                NaN   
49                                                NaN   
50               date recensement Taux de natalité(‰)   
51  date recensement Indice de fécondité(enfants /...   
52                               date recensement IPE   
53         date recensement Taux d"alphabétisation(%)   
54              date recensement Taux de mortalité(‰)   
55    date recensement Taux de mortalité infantile(‰)   
56                                PIB(milliards $USD)   
57                               date recensement PIB   
58                                       PIB/habitant   
59                      date recensement PIB/habitant   
60                               date recensement IPE   
61                                    IDH(/ 1 (2018))   
62                                  IDH(/ 1 (2018)) 2   
63                                  IDH(/ 1 (2018)) 2   
64                          Croissance du PIB(% / an)   
65         date recensement Croissance du PIB(% / an)   
66                         date recensement Touristes   
67                                       PIB/habitant   
68                          Croissance du PIB(% / an)   
69                              Espérance de vie(ans)   
70             date recensement Espérance de vie(ans)   

                                          description  \
0   Split column Population(habitants (2020)) by s...   
1   Text transform on cells in column Population(h...   
2   Rename column Population(habitants (2020)) 2 t...   
3   Text transform on cells in column Croissance d...   
4   Text transform on cells in column Superficie(k...   
5   Text transform on cells in column Densité(habi...   
6   Split column PIB(milliards $USD (2017)) by sep...   
7   Text transform on cells in column PIB(milliard..